src/main/java/io/odpf/firehose/sink/mongodb/client/MongoSinkClient.java

```diff
@@ -0,0 +1,143 @@
+ * as well as performing bulk writes to the MongoDB collection.
+ * It also logs to the instrumentation whether the bulk write has
+ * succeeded or failed, as well as the cause of the failures.
+ *
+ * @since 0.1
+ */
+@AllArgsConstructor
+public class MongoSinkClient implements Closeable {
+
+    private final MongoCollection<Document> mongoCollection;
+    private final Instrumentation instrumentation;
+    private final List<String> mongoRetryStatusCodeBlacklist;
+    private final MongoClient mongoClient;
+    private final MongoSinkConfig mongoSinkConfig;
+
+    /**
+     * Instantiates a new Mongo sink client.
+     *
+     * @param mongoSinkConfig the mongo sink config
+     * @param instrumentation the instrumentation
+     * @since 0.1
+     */
+    public MongoSinkClient(MongoSinkConfig mongoSinkConfig, Instrumentation instrumentation) {
+        this.mongoSinkConfig = mongoSinkConfig;
+        this.instrumentation = instrumentation;
+        mongoClient = MongoSinkClientUtil.buildMongoClient(mongoSinkConfig, instrumentation);
+
+        MongoDatabase database = mongoClient.getDatabase(mongoSinkConfig.getSinkMongoDBName());
+        mongoCollection = database.getCollection(mongoSinkConfig.getSinkMongoCollectionName());
+        mongoRetryStatusCodeBlacklist = MongoSinkClientUtil.getStatusCodesAsList(mongoSinkConfig.getSinkMongoRetryStatusCodeBlacklist());
+    }
+
+
+    /**
+     * Processes the bulk request list of WriteModel.
+     * This method performs a bulk write operation on the MongoCollection
+     * If bulk write succeeds, an empty list is returned
+     * If bulk write fails, then failure count is logged to instrumentation
+     * and returns a list of BulkWriteErrors, whose status codes are
+     * not present in retry status code blacklist
+     *
+     * @param request the bulk request
+     * @return the list of non-blacklisted Bulk Write errors, if any, else returns empty list
+     * @since 0.1
+     */
+    public List<BulkWriteError> processRequest(List<WriteModel<Document>> request) {
+        try {
+            logResults(mongoCollection.bulkWrite(request), request.size());
+            return Collections.emptyList();
+        } catch (MongoBulkWriteException writeException) {
+            instrumentation.logWarn("Bulk request failed");
+            List<BulkWriteError> writeErrors = writeException.getWriteErrors();
+            logErrors(writeErrors);
+            return writeErrors.stream()
+                    .filter(writeError -> !mongoRetryStatusCodeBlacklist.contains(String.valueOf(writeError.getCode())))
+                    .collect(Collectors.toList());
+        }
+    }
+
+    private void logResults(BulkWriteResult writeResult, int messageCount) {
+
+        int totalWriteCount = writeResult.getModifiedCount() + writeResult.getUpserts().size();
+        if (writeResult.wasAcknowledged()) {
+            if (mongoSinkConfig.isSinkMongoModeUpdateOnlyEnable() && totalWriteCount != messageCount) {
```

Comment:
yeah i added those checks the case when mode was UPDATE_ONLY
then when the primary key of the document to update is not found in the collection,
the message is dropped by mongo as expected but mongo library doesnt throw BulkWriteException in this case, due to which no logs are obtained for the dropped message, so ive added some code to check this

Reply:
if insert was not successful in case of upsert, then this method itself would be called, because mongo collection will throw BulkWriteException in this case, which will be handled and logged by logErrors() method